terraform/main.tf

```diff
@@ -5,155 +5,227 @@ terraform {
-    # FIXME: This is way too much perms just to ssh into a node
-    "${var.project_id}=>roles/compute.instanceAdmin.v1"
-  ]
+// Service account used by all the nodes and pods in our cluster
+resource "google_service_account" "cluster_sa" {
+  account_id   = "${var.prefix}-cluster-sa"
+  display_name = "Cluster SA for ${var.prefix}"
+  project      = var.project_id
 }
 
-output "ci_deployer_key" {
-  value     = module.service_accounts.keys["cd-sa"]
-  sensitive = true
+// To access GCS buckets with requestor pays, the calling code needs
+// to have serviceusage.services.use permission. We create a role
+// granting just this to provide the cluster SA, so user pods can
+// use it. See https://cloud.google.com/storage/docs/requester-pays
+// for more info
+resource "google_project_iam_custom_role" "identify_project_role" {
+  // Role names can't contain -, so we swap them out. BOO
+  role_id     = replace("${var.prefix}_user_sa_role", "-", "_")
+  project     = var.project_id
+  title       = "Identify as project role for users in ${var.prefix}"
+  description = "A description"
+  permissions = ["serviceusage.services.use"]
 }
 
-resource "google_artifact_registry_repository" "container_repository" {
-  provider = google-beta
-
-  location      = var.region
-  repository_id = "low-touch-hubs"
-  format        = "DOCKER"
-  project       = var.project_id
+resource "google_project_iam_member" "identify_project_binding" {
+  project = var.project_id
+  role    = google_project_iam_custom_role.identify_project_role.name
+  member  = "serviceAccount:${google_service_account.cluster_sa.email}"
 }
 
-// Give the GKE service account access to our artifact registry docker repo
-resource "google_project_iam_member" "project" {
+resource "google_project_iam_member" "cluster_sa_roles" {
+  for_each = var.cluster_sa_roles
+
   project = var.project_id
-  role    = "roles/artifactregistry.reader"
-  member  = "serviceAccount:${module.gke.service_account}"
+  role    = each.value
+  member  = "serviceAccount:${google_service_account.cluster_sa.email}"
 }
 
+resource "google_container_cluster" "cluster" {
+  # config_connector_config is in beta
+  provider = google-beta
+
+  name     = "${var.prefix}-cluster"
+  location = var.zone
+  project  = var.project_id
 
-module "gke" {
-  source                     = "terraform-google-modules/kubernetes-engine/google"
-  project_id                 = var.project_id
-  name                       = "${var.prefix}-cluster"
-  regional                   = var.regional_cluster
-  region                     = var.region
-  zones                      = [var.zone]
-  network                    = "default"
-  subnetwork                 = "default"
-  ip_range_pods              = ""
-  ip_range_services          = ""
-  http_load_balancing        = false
-  horizontal_pod_autoscaling = false
-  network_policy             = true
-  # We explicitly set up a core pool, so don't need the default
+  initial_node_count       = 1
```

Comment:
So you create a node pool and then you delete it with `remove_default_node_pool = true`.
This is because you can create a cluster with zero node pools, right?

Reply:
yep, it's because we can't create a cluster with zero nodepools...